# Mushroom Prediction: A Preliminary Notebook

<div style="border: 2px double #dcdcdc; padding: 10px; border-radius: 5px; background-color: #202020; max-width: 97.5%; overflow-x: auto;">
<h2> Version Log </h2>
<p>
<br>[3_8_2024]: Completed data preprocessing. Constructed and trained various base-lv models and ensembled into a meta-model. Made prediction, MCC = 0.98

<br>[2_8_2024]: Completed data cleaning 
</p>
</div>

## 1. Setup Environment:

In [40]:
## This is a Juptyer notebook for the Kaggle Project: Mushroom Classification
# %pip install ydata-profiling
# %pip install numpy
# %pip install --upgrade pandas
# %pip install --upgrade matplotlib
# %pip install --upgrade seaborn
# %pip install --upgrade scikit-learn
# %pip install --upgrade scipy
# %pip install --upgrade catboost
# %pip install --upgrade xgboost
# %pip install --upgrade lightgbm
# %pip install ipywidgets

In [41]:
## Import libaries
## Data analysis and wrangling
import numpy as np
import pandas as pd
import random as rnd

## Data Visualization
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import set_config
from ydata_profiling import ProfileReport
%matplotlib inline 
from scipy.stats import boxcox

# Metrics
# from sklearn.metrics import mean_squared_error
# from sklearn.metrics import mean_absolute_error
# from sklearn.metrics import r2_score
# from sklearn.metrics import mean_squared_log_error
from sklearn.metrics import make_scorer
from sklearn.metrics import matthews_corrcoef
# from sklearn.metrics import root_mean_squared_log_error

# # Machine learning_ Regression
# from sklearn.linear_model import LinearRegression, LogisticRegression, Ridge, Lasso, ElasticNet,Perceptron, SGDClassifier
# from sklearn.svm import SVR
# from sklearn.ensemble import  AdaBoostRegressor, RandomForestRegressor, GradientBoostingRegressor, ExtraTreesRegressor
# from sklearn.neighbors import KNeighborsRegressor
# from sklearn.naive_bayes import GaussianNB
# from catboost import CatBoostRegressor
# from xgboost import XGBRegressor
# from sklearn.tree import DecisionTreeRegressor
# import lightgbm as lgb
# from sklearn.neural_network import MLPRegressor
# from sklearn.linear_model import LogisticRegression
# from sklearn.dummy import DummyRegressor

# Machine learning_ Classification
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.tree import DecisionTreeClassifier, ExtraTreeClassifier
from sklearn.svm import SVC, LinearSVC
from xgboost import XGBClassifier
from catboost import CatBoostClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.dummy import DummyClassifier

# # Model selection
from sklearn.model_selection import cross_val_score, cross_val_predict
from sklearn.model_selection import KFold, StratifiedKFold, train_test_split
from sklearn.model_selection import GridSearchCV, RandomizedSearchCV


#Palette
palette = ['#328ca9', '#0e6ea9', '#2c4ea3', '#193882', '#102446']

# Set the style of the visualization
sns.set(style="whitegrid")

# Set the configuration of sklearn
SEED = 42 # for reproducibility

## 2.  Problem identification

### Problem Statement:

This is one of the 2024 playground competitions on Kaggle. 

The major goal of the project is to develope a classifier for classifying muchrooms into edible or poisonous based on physical characteristics presented in tabulated formats. 
The performance of the model will be assessed by the Matthews correlation coefficient (MCC), that is calculated by:

$$MCC = \frac{TP \cdot TN - FP \cdot FN}{\sqrt{(TP + FP)(TP + FN)(TN + FP)(TN + FN)}}$$


<br> Reference:
<br> [1] Walter Reade, Ashley Chow. (2024). Binary Prediction of Poisonous Mushrooms. Kaggle. https://kaggle.com/competitions/playground-series-s4e8
<br> [2] https://archive.ics.uci.edu/dataset/73/mushroom 

## 3. Reading Data

In [42]:
## Reading data
train_df=pd.read_csv(r'Input\train.csv')#
test_df=pd.read_csv(r'Input\test.csv')

## 4. Data Inspection

### 4.1. The number of features

In [43]:
# Number of columns and rows in the dataset
print(train_df.columns.values)
print(train_df.shape)

['id' 'class' 'cap-diameter' 'cap-shape' 'cap-surface' 'cap-color'
 'does-bruise-or-bleed' 'gill-attachment' 'gill-spacing' 'gill-color'
 'stem-height' 'stem-width' 'stem-root' 'stem-surface' 'stem-color'
 'veil-type' 'veil-color' 'has-ring' 'ring-type' 'spore-print-color'
 'habitat' 'season']
(3116945, 22)


<div style="border: 2px solid #000000; padding: 10px; border-radius: 5px; background-color: #282828; max-width: 97.5%; overflow-x: auto;">
<p>
<br> There are 22 features in the 3116945 entries in the training dataset. 
<br> The id is the index of the data entry; the class is the target variable and the others are features.  
<br> The cap, bruise, gill, stem, veil and ring are different parts of a mushroom that can be found in the anatomy shown below. </p>
</div>

From https://archive.ics.uci.edu/dataset/73/mushroom, the features included are:
1. id                   = the index of the data entry
2. class                = e (edible)/ p (poisonous) - <b>The target variable </b>
3. cap-diameter         = The di
4. cap-shape            = a shape descriptor of the mushroom's cap 
5. cap-surface          = a surface descriptor 
6. cap-color            = the color of the mushroom cap
7. does-bruise-or-bleed = Will the mushroom change color when being bruised or cut 
8. gill-attachment      = Does the mushroom have gill attached
9. gill-spacing         = the spacing of gill under the cap of mushroom  
10. gill-color          = the color of the gill
11. stem-height         = 
12. stem-width          = 
13. stem-root           = 
14. stem-surface        = 
15. stem-color          = the color of the stem
16. veil-type           = the type of the veil 
17. veil-color          = the color of the veil
18. has-ring            = the existance of rings
19. ring-type           = the type of rings.
20. spore-print-color   = the color of print of spore obtained by cutting the cap and gill and cover with a blank paper.
21. habitat             = the habit of the mushrooms.
22. season              = the season of obtaining the mushrooms.



![alt text](<Anatomy of a Mushroom Graphic-1.webp>)

In [44]:
# Look at the first 5 rows of the dataset
train_df.head()

id class  cap-diameter cap-shape cap-surface cap-color  \
0   0     e          8.80         f           s         u   
1   1     p          4.51         x           h         o   
2   2     e          6.94         f           s         b   
3   3     e          3.88         f           y         g   
4   4     e          5.85         x           l         w   

  does-bruise-or-bleed gill-attachment gill-spacing gill-color  ...  \
0                    f               a            c          w  ...   
1                    f               a            c          n  ...   
2                    f               x            c          w  ...   
3                    f               s          NaN          g  ...   
4                    f               d          NaN          w  ...   

   stem-root  stem-surface stem-color veil-type veil-color has-ring ring-type  \
0        NaN           NaN          w       NaN        NaN        f         f   
1        NaN             y          o       NaN        NaN        t         z   
2        NaN             s          n       NaN        NaN        f         f   
3        NaN           NaN          w       NaN        NaN        f         f   
4        NaN           NaN          w       NaN        NaN        f         f   

  spore-print-color habitat season  
0               NaN       d      a  
1               NaN       d      w  
2               NaN       l      w  
3               NaN       d      u  
4               NaN       g      a  

[5 rows x 22 columns]

In [45]:
# Look at the last 5 rows of the dataset
train_df.tail()

id class  cap-diameter cap-shape cap-surface cap-color  \
3116940  3116940     e          9.29         f         NaN         n   
3116941  3116941     e         10.88         s         NaN         w   
3116942  3116942     p          7.82         x           e         e   
3116943  3116943     e          9.45         p           i         n   
3116944  3116944     p          3.20         x           s         g   

        does-bruise-or-bleed gill-attachment gill-spacing gill-color  ...  \
3116940                    t             NaN          NaN          w  ...   
3116941                    t               d            c          p  ...   
3116942                    f               a          NaN          w  ...   
3116943                    t               e          NaN          p  ...   
3116944                    f               d            c          w  ...   

         stem-root  stem-surface stem-color veil-type veil-color has-ring  \
3116940          b           NaN          w         u          w        t   
3116941        NaN           NaN          w       NaN        NaN        f   
3116942        NaN           NaN          y       NaN          w        t   
3116943        NaN             y          w       NaN        NaN        t   
3116944        NaN           NaN          w       NaN        NaN        f   

        ring-type spore-print-color habitat season  
3116940         g               NaN       d      u  
3116941         f               NaN       d      u  
3116942         z               NaN       d      a  
3116943         p               NaN       d      u  
3116944         f               NaN       g      u  

[5 rows x 22 columns]

### 4.2. Data type of the variables

In [46]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3116945 entries, 0 to 3116944
Data columns (total 22 columns):
 #   Column                Dtype  
---  ------                -----  
 0   id                    int64  
 1   class                 object 
 2   cap-diameter          float64
 3   cap-shape             object 
 4   cap-surface           object 
 5   cap-color             object 
 6   does-bruise-or-bleed  object 
 7   gill-attachment       object 
 8   gill-spacing          object 
 9   gill-color            object 
 10  stem-height           float64
 11  stem-width            float64
 12  stem-root             object 
 13  stem-surface          object 
 14  stem-color            object 
 15  veil-type             object 
 16  veil-color            object 
 17  has-ring              object 
 18  ring-type             object 
 19  spore-print-color     object 
 20  habitat               object 
 21  season                object 
dtypes: float64(3), int64(1), object(18)
memory

<div style="border: 2px solid #000000; padding: 10px; border-radius: 5px; background-color: #282828; max-width: 97.5%; overflow-x: auto;">
<p>
<br> It is observed that most of the variables are categoiral, except id, cap-diameter , stem-height and stem-width, which are numerical.  </p>
</div>


In [47]:
# Display the unique values and the count of unique values in the dataset
train_df.describe(include=['O'])


class cap-shape cap-surface cap-color does-bruise-or-bleed  \
count   3116945   3116905     2445922   3116933              3116937   
unique        2        74          83        78                   26   
top           p         x           t         n                    f   
freq    1705396   1436026      460777   1359542              2569743   

       gill-attachment gill-spacing gill-color stem-root stem-surface  \
count          2593009      1858510    3116888    359922      1136084   
unique              78           48         63        38           60   
top                  a            c          w         b            s   
freq            646034      1331054     931538    165801       327610   

       stem-color veil-type veil-color has-ring ring-type spore-print-color  \
count     3116907    159452     375998  3116921   2988065            267263   
unique         59        22         24       23        40                32   
top             w         u          w        f         f                 k   
freq      1196637    159373     279070  2368820   2477170            107310   

        habitat   season  
count   3116900  3116945  
unique       52        4  
top           d        a  
freq    2177573  1543321

### 4.3 Handling missing, distinct and duplicated entries:

#### 4.3.1 Number of Missing/ distinct and duplicated entries:

In [48]:
print("Missing Data in the training dataset (%)")
print("="*40)
print((train_df.isnull().sum()/len(train_df)*100).sort_values(ascending = False).round(3))
print("="*40)

Missing Data in the training dataset (%)
veil-type               94.884
spore-print-color       91.425
stem-root               88.453
veil-color              87.937
stem-surface            63.551
gill-spacing            40.374
cap-surface             21.528
gill-attachment         16.809
ring-type                4.135
gill-color               0.002
habitat                  0.001
cap-shape                0.001
stem-color               0.001
has-ring                 0.001
cap-color                0.000
does-bruise-or-bleed     0.000
cap-diameter             0.000
id                       0.000
stem-width               0.000
class                    0.000
stem-height              0.000
season                   0.000
dtype: float64


In [49]:
print("Distinct Data in the training dataset (%)")
print("="*40)
print(train_df.nunique().sort_values(ascending = False))
print("="*40)

Distinct Data in the training dataset (%)
id                      100.000
stem-width                0.187
cap-diameter              0.126
stem-height               0.088
cap-surface               0.003
cap-color                 0.003
gill-attachment           0.003
cap-shape                 0.002
gill-color                0.002
stem-surface              0.002
stem-color                0.002
habitat                   0.002
gill-spacing              0.002
ring-type                 0.001
stem-root                 0.001
spore-print-color         0.001
does-bruise-or-bleed      0.001
veil-color                0.001
has-ring                  0.001
veil-type                 0.001
season                    0.000
class                     0.000
dtype: float64


In [50]:
print("Duplicated Data in the training dataset (%)")
print("="*40)
print((train_df.duplicated().sum()/len(train_df)*100).round(3))
print("="*40)

Duplicated Data in the training dataset (%)
0.0


<div style="border: 2px solid #000000; padding: 10px; border-radius: 5px; background-color: #282828; max-width: 97.5%; overflow-x: auto;">
<p>
<br>  There are multiple missing values in the data with some of them contribute a large portion of the entries (e.g. veil-type & spore-print-color). </p>
</div>



b. Understand the cause and type of missing data. 

Different types of missing data may leads to different handling methods. Following are some examples of missing data and their handling methods.

Missing Completely at Random (MCAR):

Handling method: Simple imputation
Cause: The probability of missing an entry is completely independent of the values of the variables in the dataset, as well as the unobserved data.
Example: a sensor malfunctions and randomly fails to record some measurements
Missing at Random (MAR):

Handling method: advanced imputation techniques
Cause: The probability of a data point being missing depends on the observed variables in the dataset, but not on the unobserved (missing) data.
Example: income data is more likely to be missing for individuals with lower education levels
Missing Not at Random (MNAR):

Handling methods: pattern mixture models or selection models
Causes: the probability of a data point being missing depends on the unobserved (missing) data itself.
Example: individuals with higher income are less likely to report their income
Systematic Missing Data:

Handling methods: Simple imputation or creating Proxy variable
Causes: An entire variables or features are missing from the dataset, typically due to issues in data collection or data processing.
Example: A sensor was not installed on certain devices, resulting in the absence of data for a specific feature

> Right now, we will choose to impute the missing values by the most frequent option in the categorical features. And for distint values, we will keep only the most frequent 5 options in each feature to reduce the cardinality. 

c. Handling of the missing/distinct/diplicated data.



In [51]:
# Drop features that have > 50% missing values
train_df = train_df.dropna(thresh=0.5*len(train_df), axis=1)
# Drop the 'target' column and assign it to the target variable
y = train_df['class']
train_df = train_df.drop(['class'], axis=1)
# Drop the same features in the test dataset
test_df = test_df[train_df.columns]


In [52]:
# imputate missing values
for col in train_df.columns:
    if train_df[col].isnull().sum() > 0:
        # print(f"Imputing {col} with mode")
        train_df[col].fillna(train_df[col].mode()[0], inplace=True)

for col in test_df.columns:
    if test_df[col].isnull().sum() > 0:
        # print(f"Imputing {col} with mode")
        test_df[col].fillna(test_df[col].mode()[0], inplace=True)
# Check if there are any missing values
print("Missing Data in the training dataset (%)")
print("="*40)
print((train_df.isnull().sum()/len(train_df)*100).sort_values(ascending = False).round(3))
print("="*40)


Imputing cap-diameter with mode
Imputing cap-shape with mode


C:\Users\Kai Qi Yan\AppData\Local\Temp\ipykernel_29884\3961555843.py:5: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  train_df[col].fillna(train_df[col].mode()[0], inplace=True)


Imputing cap-surface with mode
Imputing cap-color with mode
Imputing does-bruise-or-bleed with mode
Imputing gill-attachment with mode
Imputing gill-spacing with mode
Imputing gill-color with mode
Imputing stem-color with mode
Imputing has-ring with mode
Imputing ring-type with mode
Imputing habitat with mode
Imputing cap-diameter with mode
Imputing cap-shape with mode
Imputing cap-surface with mode


C:\Users\Kai Qi Yan\AppData\Local\Temp\ipykernel_29884\3961555843.py:10: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  test_df[col].fillna(test_df[col].mode()[0], inplace=True)


Imputing cap-color with mode
Imputing does-bruise-or-bleed with mode
Imputing gill-attachment with mode
Imputing gill-spacing with mode
Imputing gill-color with mode
Imputing stem-height with mode
Imputing stem-color with mode
Imputing has-ring with mode
Imputing ring-type with mode
Imputing habitat with mode
Missing Data in the training dataset (%)
id                      0.0
cap-diameter            0.0
cap-shape               0.0
cap-surface             0.0
cap-color               0.0
does-bruise-or-bleed    0.0
gill-attachment         0.0
gill-spacing            0.0
gill-color              0.0
stem-height             0.0
stem-width              0.0
stem-color              0.0
has-ring                0.0
ring-type               0.0
habitat                 0.0
season                  0.0
dtype: float64


Next we are going to keep only the 5 most frequent values in the categorial columns, others will be categorized as others. 

In [53]:
# Drop the 'Id' colum since it's unnecessary for the prediction process
train_df = train_df.drop(['id'], axis=1)
test_df = test_df.drop(['id'], axis=1)

# separate the numerical and categorical columns
numerical_cols = train_df.select_dtypes(include=[np.number]).columns.tolist()
categorical_cols = train_df.select_dtypes(include=[object]).columns.tolist()
# Keep only the 5 most frequent values in the categorical columns
for col in categorical_cols:
    top_5 = train_df[col].value_counts().head(5).index
    train_df.loc[~train_df[col].isin(top_5), col] = 'Other'
    test_df.loc[~test_df[col].isin(top_5), col] = 'Other'

train_df

cap-diameter cap-shape cap-surface cap-color does-bruise-or-bleed  \
0                8.80         f           s     Other                    f   
1                4.51         x           h     Other                    f   
2                6.94         f           s     Other                    f   
3                3.88         f           y         g                    f   
4                5.85         x       Other         w                    f   
...               ...       ...         ...       ...                  ...   
3116940          9.29         f           t         n                    t   
3116941         10.88         s           t         w                    t   
3116942          7.82         x       Other         e                    f   
3116943          9.45     Other       Other         n                    t   
3116944          3.20         x           s         g                    f   

        gill-attachment gill-spacing gill-color  stem-height  stem-width  \
0                     a            c          w         4.51       15.39   
1                     a            c          n         4.79        6.48   
2                     x            c          w         6.85        9.93   
3                     s            c          g         4.16        6.53   
4                     d            c          w         3.37        8.36   
...                 ...          ...        ...          ...         ...   
3116940               a            c          w        12.14       18.81   
3116941               d            c          p         6.65       26.97   
3116942               a            c          w         9.51       11.06   
3116943               e            c          p         9.13       17.77   
3116944               d            c          w         2.82        7.79   

        stem-color has-ring ring-type habitat season  
0                w        f         f       d      a  
1                o        t         z       d      w  
2                n        f         f       l      w  
3                w        f         f       d      u  
4                w        f         f       g      a  
...            ...      ...       ...     ...    ...  
3116940          w        t     Other       d      u  
3116941          w        f         f       d      u  
3116942          y        t         z       d      a  
3116943          w        t     Other       d      u  
3116944          w        f         f       g      u  

[3116945 rows x 15 columns]

4.4. Summary of the dataset by ProfileReport

In [54]:
# profile=ProfileReport(train_df,title='Pandas Profiling Report',explorative=True)
# profile.to_notebook_iframe()

## 5. Feature Preprocessing

In [55]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.preprocessing import LabelEncoder
# visualizing pipeline
set_config(display='diagram')

# Define a function to preprocess the data
def PreprocessData(df):

    # Define the numerical and categorical features
    num_features = df._get_numeric_data().columns
    cat_features = list(set(df.columns) - set(num_features))

    # Set up the numerical and categorical transformers
    numeric_transformer = Pipeline(
    steps=[("imputation_mean",SimpleImputer(missing_values=np.nan,strategy="mean")),
          ("scaler",StandardScaler())]
    )
    categorial_transformer = Pipeline(
    steps=[("imputation_mode",SimpleImputer(missing_values=np.nan,strategy="most_frequent")),('onehot',OneHotEncoder(handle_unknown='ignore'))]
    )
    # Set up the preprocessor
    preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, num_features),
        ('cat', categorial_transformer, cat_features)
    ])

    # Fit and transform the data
    X = preprocessor.fit_transform(df)

    # Return the preprocessed data
    return X, preprocessor

In [56]:
# Preprocess the data
X, preprocessor = PreprocessData(train_df)
X1, preprocessor = PreprocessData(test_df)
# Encode the target variable
le = LabelEncoder()
y = le.fit_transform(y)


## 6. 1st Level models

### 6.1. Model list

In [57]:
# create a utility function to sort the dictionary by values aphabatically
def sort_dict(d):
    return dict(sorted(d.items(), key=lambda x: x[0]))

# Define a list of models for prediction
classifiers = {
    "Logistic Regression": LogisticRegression(),
    # "Random Forest Classifier": RandomForestClassifier(),
    # "Gradient Boosting Classifier": GradientBoostingClassifier(),
#     "Support Vector Classifier": SVC(),
    "XGBClassifier": XGBClassifier(),
#     "CatBoost Classifier": CatBoostClassifier(verbose=False),
#     "MLP Classifier": MLPClassifier(),
#     "Extra Tree Classifier": ExtraTreeClassifier(),
    "Dummy Classifier": DummyClassifier(strategy='most_frequent')  # DummyClassifier for sanity check
}

# Sort the models
classifiers=sort_dict(classifiers)

### 6.2 Model parameters

In [58]:
# Define the parameters for the hyperparamter tuning of models by RandomizedSearchCV
import scipy.stats as stats

params_classifiers = {

    "Logistic Regression": {
        'solver': ['newton-cg', 'sag', 'lbfgs'],  
        'penalty': ['l2', None],  
        'C': [0.1, 1, 10]
    },

    # "Random Forest Classifier": {
    #     'n_estimators': [8, 16, 32, 64, 128, 256],
    #     'criterion': ['gini', 'entropy'],
    #     'warm_start': [True]

    # },

    # "Gradient Boosting Classifier": {
    #     'learning_rate': stats.loguniform(1e-2, 1e-1),
    #     'n_estimators': [8, 16, 32, 64, 128, 256],
    #     'subsample': [0.6, 0.7, 0.75, 0.8, 0.85, 0.9],
    #     'max_features': ['auto', 'sqrt', 'log2']
    # },

#     "Support Vector Classifier": {
#         'C': stats.loguniform(1e-2, 1e2),
#         'kernel': ['linear', 'poly', 'rbf', 'sigmoid'],
#         'gamma': ['scale', 'auto']
#     },

    "XGBClassifier": {
            'objective':['binary:logistic'],
            'error_score':['raise'],
            'max_depth': [3, 5, 7, 9],
            'colsample_bytree': [0.6, 0.7, 0.8, 0.9],
            'n_estimators': [16, 32, 64, 128, 256]
        },


#     "CatBoost Classifier": {
#         'depth': [6, 8, 10],
#         'learning_rate': stats.loguniform(1e-2, 1e-1),
#         'iterations': [30, 50, 100]
#     },

#     "MLP Classifier": {
#         'hidden_layer_sizes': [(50, 50, 50), (50, 100, 50), (100,)],
#         'activation': ['tanh', 'relu'],
#         'solver': ['sgd', 'adam'],
#         'alpha': stats.loguniform(1e-5, 1e-2),
#         'learning_rate': ['constant', 'adaptive']
#     },
    
#     "Extra Tree Classifier": {
#         'n_estimators': [8, 16, 32, 64, 128, 256],
#         'criterion': ['gini', 'entropy']
#     },

    "Dummy Classifier": {}
}

# Sort the parameters
params_classifiers = sort_dict(params_classifiers)

6.3. Model training and hyperparameters tuning

a. Setup scoring method for the model optimization

In [59]:
from sklearn.model_selection import StratifiedKFold

# Setup the KFold
NFOLDS = 3 # set folds for out-of-fold prediction
kf = StratifiedKFold(n_splits= NFOLDS,shuffle=True, random_state=SEED)

In [60]:
from sklearn.metrics import make_scorer
from sklearn.metrics import matthews_corrcoef

# Define a custom cost function for the RandomizedSearchCV
def MCC(y_true, y_pred):
    MCC= matthews_corrcoef(y_true, y_pred)
    return MCC

In [61]:
# Define a function to obtain the out of fold predictions
def get_oof_predictions(model, X_train, y_train, X_test, kf):
    
    # Initialize the out of fold predictions
    oof_predictions = np.zeros((X_train.shape[0],))
    # Initialize the test predictions
    test_predictions = np.zeros((X_test.shape[0],))

    # Loop through the training and validation sets
    for train_index, val_index in kf.split(X_train,y_train):
        # Split the training and validation sets
        X_tr, X_val = X_train[train_index], X_train[val_index]
        y_tr, y_val = y_train[train_index], y_train[val_index]
        # Fit the model
        model.fit(X_tr, y_tr) # is this necessary?
        # Make predictions
        oof_predictions[val_index] = model.predict(X_val)
        test_predictions += model.predict(X_test)

    # Return the out of fold predictions and the test predictions
    return model,oof_predictions, test_predictions / kf.get_n_splits() # Average the test predictions

In [62]:
import time
from joblib import dump, load
import os
import numpy as np
from sklearn.model_selection import RandomizedSearchCV
from sklearn.metrics import make_scorer


def model_evaluation(models, X_train, y_train, X_val, y_val, kf,params=None,mode='tuning'):
    model_list = []
    MCC_train_list = []
    MCC_val_list = []
    time_list = []
    y_train_pred_list = []
    model_params = []
    # Initialize a dataframe of out of fold predictions for the models
    oof_predictions_df = pd.DataFrame()
    # Initialize a dataframe of test predictions for the models
    test_predictions_df = pd.DataFrame()

    for i in range(len(list(models))):
        # check if the model is saved in the dictionary
        if os.path.exists(f"Output/{list(models.keys())[i]}"+"_"+mode+".joblib"):
            model = load(f"Output/{list(models.keys())[i]}"+"_"+mode+".joblib")
            y_train_pred = model.predict(X_train)
            y_test_pred = model.predict(X_val)

            MCC_train = MCC(y_train, y_train_pred)
            MCC_test = MCC(y_val, y_test_pred)

            model_list.append(list(models.keys())[i])
            MCC_train_list.append(MCC_train)
            MCC_val_list.append(MCC_test)
            y_train_pred_list.append(y_train_pred)
            model_params.append(model.get_params())

            print('Model Success:', list(models.keys())[i], 'MCC_train:', MCC_train, ' , MCC_test:', MCC_test)
            print('=' * 35)
            print('\n')
        else:
            try:
                model = list(models.values())[i]
                para = params[list(models.keys())[i]]
                
                if mode == 'tuning':
                    RS = RandomizedSearchCV(model, para, n_iter=10, scoring=make_scorer(MCC, greater_is_better=True), refit=True, n_jobs=6, verbose=1)
                    RS.fit(X_train, y_train)
                    model.set_params(**RS.best_params_)
                    model.fit(X_train, y_train)
                    best_model = RS.best_estimator_
                    model_params.append(RS.best_params_)

                elif mode == 'training':
                    # Get the out of fold predictions and the test predictions
                    best_model,oof_predictions, test_predictions = get_oof_predictions(model, X_train, y_train, X_val, kf)
                    # Add the out of fold predictions to the dataframe
                    oof_predictions_df[list(models.keys())[i]] = oof_predictions
                    # Add the test predictions to the dataframe
                    test_predictions_df[list(models.keys())[i]] = test_predictions

                # Make predictions
                y_train_pred = best_model.predict(X_train)
                y_test_pred = best_model.predict(X_val)

                # Evaluate Train and Test dataset
                MCC_train = MCC(y_train, y_train_pred)
                MCC_test = MCC(y_val, y_test_pred)

                # Append the results to the lists
                model_list.append(list(models.keys())[i])
                MCC_train_list.append(MCC_train)
                MCC_val_list.append(MCC_test)
                y_train_pred_list.append(y_train_pred)

                # Save the best parameters in a directory with the name of the model
                dump(best_model, f"Output/{list(models.keys())[i]}"+"_"+mode+".joblib", protocol=5)
                
                print('Model Success:', list(models.keys())[i], 'MCC_train:', MCC_train, ' , MCC_test:', MCC_test)
                print('=' * 35)
                print('\n')

            # Raise exception if the model fails
            except Exception as e:
                print(list(models.keys())[i])
                model_list.append(list(models.keys())[i])
                MCC_train_list.append(np.nan)
                MCC_val_list.append(np.nan)
                y_train_pred_list.append(np.nan)
                model_params.append(np.nan)
                print('Model failed:', e)
                print('=' * 35)
                print('\n')
                continue

    # save the results in a dataframe
    results = pd.DataFrame({'Model': model_list, 'MCC_train': MCC_train_list, 'MCC_test': MCC_val_list})
    # save the results in a csv file
    results.to_csv('Output/results_'+mode+'.csv', index=False)
    if mode == 'tuning':
        return model_list, MCC_train_list, MCC_val_list, y_train_pred_list, model_params
    elif mode == 'training':
        return model_list, MCC_train_list, MCC_val_list, y_train_pred_list, oof_predictions_df, test_predictions_df

### 6.3 Hyperparameter tuning using a smaller set of data

In [63]:
# Sampling data for hyperparameter tuning
sample_size = 100000  # sample size for tuning
X_sample, _, y_sample, _ = train_test_split(X, y, train_size=sample_size, stratify=y, random_state=42)

# Split the data into training and testing sets
X_train_sample, X_val_sample, y_train_sample, y_val_sample = train_test_split(X_sample, y_sample, test_size=0.2, random_state=42, stratify=y_sample)

# MCC Scores
model_list_tuning, MCC_train_list_tuning, MCC_val_list_tuning,y_train_pred_list_tuning, model_params_tuning = model_evaluation(classifiers, X_train_sample, y_train_sample, X_val_sample, y_val_sample, kf, params= params_classifiers, mode='tuning')

# Display the scores
pd.DataFrame(list(zip(model_list_tuning, MCC_train_list_tuning, MCC_val_list_tuning)), columns=['Model Name', 'MCC_Score_Train_sample', 'MCC_Score_Test_sample']).sort_values(by=["MCC_Score_Train_sample"],ascending=True)

Model Success: Dummy Classifier MCC_train: 0.0  , MCC_test: 0.0


Model Success: Logistic Regression MCC_train: 0.4818733884073294  , MCC_test: 0.4718180971098021


Model Success: XGBClassifier MCC_train: 0.9863354598812126  , MCC_test: 0.9777040250135468




Model Name  MCC_Score_Train_sample  MCC_Score_Test_sample
0     Dummy Classifier                0.000000               0.000000
1  Logistic Regression                0.481873               0.471818
2        XGBClassifier                0.986335               0.977704

### 6.4 Train the models with the best parameters with the entire dataset and produce Out of Fold Predictions

In [64]:
# Split the data into training and validation sets
X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2, random_state=42, stratify=y)

## Train the models with the best parameters with the entire dataset

# Add the best parameters to the models
for i in range(len(list(classifiers))):
    classifiers[list(classifiers.keys())[i]].set_params(**model_params_tuning[i])

# Train the models with the entire dataset
model_list_training, MCC_train_list_training, MCC_val_list_training, y_train_pred_list_training, oof_predictions_df, val_predictions_df = model_evaluation(classifiers, X_train, y_train, X_val, y_val, kf, params=params_classifiers, mode='training')

Model Success: Dummy Classifier MCC_train: 0.0  , MCC_test: 0.0


Model Success: Logistic Regression MCC_train: 0.4814379924453498  , MCC_test: 0.4811767113401799




c:\Users\Kai Qi Yan\anaconda3\envs\MLEnv\Lib\site-packages\xgboost\core.py:158: UserWarning: [01:35:10] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-0015a694724fa8361-1\xgboost\xgboost-ci-windows\src\learner.cc:740: 
Parameters: { "error_score" } are not used.

  warnings.warn(smsg, UserWarning)
c:\Users\Kai Qi Yan\anaconda3\envs\MLEnv\Lib\site-packages\xgboost\core.py:158: UserWarning: [01:35:28] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-0015a694724fa8361-1\xgboost\xgboost-ci-windows\src\learner.cc:740: 
Parameters: { "error_score" } are not used.

  warnings.warn(smsg, UserWarning)
c:\Users\Kai Qi Yan\anaconda3\envs\MLEnv\Lib\site-packages\xgboost\core.py:158: UserWarning: [01:35:47] WARNING: C:\buildkite-agent\builds\buildkite-windows-cpu-autoscaling-group-i-0015a694724fa8361-1\xgboost\xgboost-ci-windows\src\learner.cc:740: 
Parameters: { "error_score" } are not used.

  warnings.warn(smsg, UserWarning)


Model Success: XGBClassifier MCC_train: 0.9812258630223114  , MCC_test: 0.9798944222933721




In [65]:
# Use the trained models to make predictions on the test dataset
# Load the models
model_list_training = []
for i in range(len(list(classifiers))):
    model = load(f"Output/{list(classifiers.keys())[i]}_training.joblib")
    model_list_training.append(model)

# Make predictions on the test dataset
test_predictions_df = pd.DataFrame()
for i in range(len(model_list_training)):
    test_predictions_df[list(classifiers.keys())[i]] = model_list_training[i].predict(X1)

In [68]:
# Create a dataframe to store the OOF predictions from selected models
oof_predictions_df_selected = pd.DataFrame()
oof_predictions_df_selected = oof_predictions_df[['Logistic Regression','XGBClassifier']].copy()
val_predictions_df_selected = pd.DataFrame()
val_predictions_df_selected = val_predictions_df[['Logistic Regression','XGBClassifier']].copy()
test_predictions_df_selected = pd.DataFrame()
test_predictions_df_selected = test_predictions_df[['Logistic Regression','XGBClassifier']].copy()


## 7. Meta-model via MLPRegressor

In [69]:
## Train a meta model with the OOF predictions
# Define a MLP Classifier for the meta model
meta_model = MLPClassifier(hidden_layer_sizes=(50, 50, 50), activation='tanh', solver='adam', alpha=0.0001, learning_rate='adaptive', random_state=SEED)

# Implement a randomized search for the meta-model
# Implement a randomized search for the meta-model
params_meta = {
    'hidden_layer_sizes': [(50, 50, 50), (50, 100, 50), (100,)],
    'activation': ['tanh', 'relu'],
    'solver': ['sgd', 'adam'],
    'alpha': stats.loguniform(1e-5, 1e-2),
    'learning_rate': ['constant', 'adaptive']
}

# Setup the RandomizedSearchCV
RS_meta = RandomizedSearchCV(meta_model, params_meta, n_iter=10, cv=kf, scoring=make_scorer(MCC, greater_is_better=False),refit=True, n_jobs=-1, verbose=1)

# Fit the RandomizedSearchCV
RS_meta.fit(oof_predictions_df_selected, y_train) 

Fitting 3 folds for each of 10 candidates, totalling 30 fits


RandomizedSearchCV(cv=StratifiedKFold(n_splits=3, random_state=42, shuffle=True),
                   estimator=MLPClassifier(activation='tanh',
                                           hidden_layer_sizes=(50, 50, 50),
                                           learning_rate='adaptive',
                                           random_state=42),
                   n_jobs=-1,
                   param_distributions={'activation': ['tanh', 'relu'],
                                        'alpha': <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x000002072F4EDE90>,
                                        'hidden_layer_sizes': [(50, 50, 50),
                                                               (50, 100, 50),
                                                               (100,)],
                                        'learning_rate': ['constant',
                                                          'adaptive'],
                                        'solver': ['sgd', 'adam']},
                   scoring=make_scorer(MCC, greater_is_better=False, response_method='predict'),
                   verbose=1)

In [71]:
# Produce the final predictions from the meta-model
y_val_meta = RS_meta.predict(val_predictions_df_selected)
# Evaluate the meta-model
MCC_meta = MCC(y_val, y_val_meta)
print(f'The MCC score of the meta-model is: {MCC_meta}')

The MCC score of the meta-model is: 0.9803122384873094


In [72]:
# Obtain the best parameters of the meta-model
best_meta_model = RS_meta.best_estimator_
# Save the best meta-model
dump(best_meta_model, "Output/meta_model.joblib", protocol=5)

# Make predictions on the test dataset
# load the meta model
meta_model = load("Output/meta_model.joblib")
# Make predictions on the test dataset
test_predictions_meta = meta_model.predict(test_predictions_df_selected)

# Map predictions back to original labels
test_pred_class = le.inverse_transform(test_predictions_meta)

# Save the predictions in a csv file
submission_df = pd.read_csv('Input/sample_submission.csv')

submission_df['class'] = test_pred_class
submission_df.to_csv('Output/submission.csv', index=False)


## 9. Future Improvements

1. Create additional features using AutoML
2. Conduct Exploratory Data analysis on the data
    -  Explore the possiblity to better imputing the missing values by correlating with other features.